docs/concepts.md

```diff
@@ -40,9 +41,13 @@ This is what a typical source graph for one source file loaded by the browser ma
+These source graphs become more complex with bundles which have multiple original sources and minified original sources for which the backend will also create a pretty-printed version.
 
-When the backend sends a source location (e.g. the location of a stack frame), it uses a [`MappedLocation`](https://static.replay.io/protocol/tot/Debugger/#type-MappedLocation) which contains locations in all versions of the source.
+A recording may contain multiple sources with the same URL, this happens when the user navigated while recording or when a dev-server uses hot module replacement. If the sources have identical content we want to present these as one source to the user. To this end, the frontend identifies duplicates (called corresponding sources) of preferred and alternate sources in `ThreadFront.groupSourceIds()` by looking at their `url` and `contentHash` (it also ensures that only sources that are not original/generated versions of each other and that are both picked as a preferred or alternate source for their URL are identified as duplicates - this is probably overly cautious and the algorithm could be significantly simplified if we drop these requirements). All locations received from the backend are updated to reference the first corresponding source of the one they originally referenced. Whenever we send locations to the backend (e.g. to set a breakpoint or when running an analysis) we need to do the opposite and add locations for all corresponding sources.
+
+When the backend sends a source location (e.g. the location of a stack frame), it uses a [`MappedLocation`](https://static.replay.io/protocol/tot/Debugger/#type-MappedLocation) which contains locations in all versions of the source - one of them is the source that the browser executed and the others are original versions (from sourcemaps, pretty-printing and the extraction of inlined scripts from an html page) of it.
+
+The selection of preferred and alternate sources is implemented in `ThreadFront._chooseSourceId()`. It takes an array of sourceIds that usually come from a `MappedLocation`, removes those of minified sources (i.e. those for which a pretty-printed version exists) and inline scripts (if the corresponding html page is among the sources) and categorizes each source as "original" if it is from a sourcemap or a pretty-printed version of a source from a sourcemap and "generated" otherwise. It assumes that there is at most one "generated" source according to this categorization. If there is at least one "original" source, it picks that as the preferred source and the "generated" source (if any) as the alternate. Otherwise there should be exactly one source which is "generated" and that is the preferred source.
```

Comment:
This was a @jasonLaster suggestion, but I agree, it might be helpful to centralize a list of terms at the top of this documentation. The terms that are non-obvious IMO are:

- Original source: This is what the developer typed into their editor, before it was run through any toolchains
- Generated source: The output of running an original source through a toolchain (might be minified, transpiled, bundled, or otherwise transformed)

- Corresponding sources: a file which was loaded multiple times during a single recording, from the same URL, with the same content. For example, because the user refreshed the page.
- Alternate sources: this one is slightly tricky to express concisely, I think. It's a group of sources which map onto each other. For example, the "point" of execution might correctly be identified as having several different locations. One could be an original source (which the browser was unaware of at runtime), another could be the generated source (which the browser *did* run but which can be mapped onto the original) and another might be a pretty-printed file (which didn't exist when the browser was running, and is instead the product of us doing our own pretty-printing post-recording).
- Preferred source: A source from a group of alternate sources which is considered the "best" choice as a canonical representation of that source. Often this will be an original if one is available. If it's not, pretty-printed is preferred, and last would be the generated, potentially still minified source.
- Preferred source: the "best" (by our reckoning) source for a group of alternate sources

Reply:
I'll try to add that, the problem is that most of these terms are used with slightly different meanings in different places. But that just makes it more important to add an explanation.

Reply:
I've put expanded versions of these definitions into the PR itself, but I'm realizing now that my definitions actually conflict with how we do things right now.